In [ ]:
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Intro to Generating and Executing Python Code with Gemini 2.0

<table align="left">
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo"><br> Open in Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https:%2F%2Fraw.githubusercontent.com%2FGoogleCloudPlatform%2Fgenerative-ai%2Fmain%2Fgemini%2Fcode-execution%2Fintro_code_execution.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo"><br> Open in Colab Enterprise
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/GoogleCloudPlatform/generative-ai/main/gemini/code-execution/intro_code_execution.ipynb">
      <img src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo"><br> Open in Vertex AI Workbench
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb">
      <img width="32px" src="https://upload.wikimedia.org/wikipedia/commons/9/91/Octicons-mark-github.svg" alt="GitHub logo"><br> View on GitHub
    </a>
  </td>
</table>

<div style="clear: both;"></div>

<b>Share to:</b>

<a href="https://www.linkedin.com/sharing/share-offsite/?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="LinkedIn logo">
</a>

<a href="https://bsky.app/intent/compose?text=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="Bluesky logo">
</a>

<a href="https://twitter.com/intent/tweet?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/53/X_logo_2023_original.svg" alt="X logo">
</a>

<a href="https://reddit.com/submit?url=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit logo">
</a>

<a href="https://www.facebook.com/sharer/sharer.php?u=https%3A//github.com/GoogleCloudPlatform/generative-ai/blob/main/gemini/code-execution/intro_code_execution.ipynb" target="_blank">
  <img width="20px" src="https://upload.wikimedia.org/wikipedia/commons/5/51/Facebook_f_logo_%282019%29.svg" alt="Facebook logo">
</a>

| | |
|-|-|
| Author(s) |  [Kristopher Overholt](https://github.com/koverholt/) |

## Overview

This notebook introduces the code execution capabilities of the [Gemini 2.0 Flash model](https://cloud.google.com/vertex-ai/generative-ai/docs/gemini-v2), a new multimodal generative AI model from Google [DeepMind](https://deepmind.google/). Gemini 2.0 Flash offers improvements in speed, quality, and advanced reasoning capabilities including enhanced understanding, coding, and instruction following.

## Code Execution

A key feature of this model is [code execution](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution), which is the ability to generate and execute Python code directly within the API. If you want the API to generate and run Python code and return the results, you can use code execution as demonstrated in this notebook.

This code execution capability enables the model to generate code, execute and observe the results, correct the code if needed, and learn iteratively from the results until it produces a final output. This is particularly useful for applications that involve code-based reasoning such as solving mathematical equations or processing text.

## Objectives

In this tutorial, you will learn how to generate and execute code using the Gemini API in Vertex AI and the Google Gen AI SDK for Python with the Gemini 2.0 Flash model.

You will complete the following tasks:

- Generating and running sample Python code from text prompts
- Exploring data using code execution in multi-turn chats
- Using code execution in streaming sessions

## Getting started

### Install Google Gen AI SDK for Python


In [1]:
%pip install --upgrade --quiet google-genai

### Authenticate your notebook environment (Colab only)

If you're running this notebook on Google Colab, run the cell below to authenticate your environment.

In [2]:
import sys

if "google.colab" in sys.modules:
    from google.colab import auth

    auth.authenticate_user()

### Import libraries

In [3]:
import os

from IPython.display import Markdown, display
from google import genai
from google.genai.types import GenerateContentConfig, Tool, ToolCodeExecution

### Connect to a generative AI API service

Google Gen AI APIs and models including Gemini are available in the following two API services:

- [Google AI for Developers](https://ai.google.dev/gemini-api/docs): Experiment, prototype, and deploy small projects.
- [Vertex AI](https://cloud.google.com/vertex-ai/generative-ai/docs/overview): Build enterprise-ready projects on Google Cloud.
The Google Gen AI SDK provides a unified interface to these two API services.

This notebook shows how to use the Google Gen AI SDK with the Gemini API in Vertex AI.

### Set Google Cloud project information and create client

To get started using Vertex AI, you must have an existing Google Cloud project and [enable the Vertex AI API](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com).

Learn more about [setting up a project and a development environment](https://cloud.google.com/vertex-ai/docs/start/cloud-environment).

In [4]:
PROJECT_ID = "[your-project-id]"  # @param {type: "string", placeholder: "[your-project-id]", isTemplate: true}
if not PROJECT_ID or PROJECT_ID == "[your-project-id]":
    PROJECT_ID = str(os.environ.get("GOOGLE_CLOUD_PROJECT"))

LOCATION = os.environ.get("GOOGLE_CLOUD_REGION", "us-central1")

In [5]:
client = genai.Client(vertexai=True, project=PROJECT_ID, location=LOCATION)

## Working with code execution in Gemini 2.0

### Load the Gemini model

The following code loads the Gemini 2.0 Flash model. You can learn about all Gemini models on Vertex AI by visiting the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/learn/models):

In [6]:
MODEL_ID = "gemini-2.0-flash-exp"  # @param {type: "string"}

### Define the code execution tool

The following code initializes the code execution tool by passing `code_execution` in a `Tool` definition.

Later we'll register this tool with the model that it can use to generate and run Python code:

In [7]:
code_execution_tool = Tool(code_execution=ToolCodeExecution())

### Generate and execute code

The following code sends a prompt to the Gemini model, asking it to generate and execute Python code to calculate the sum of the first 50 prime numbers. The code execution tool is passed in so the model can generate and run the code:

In [8]:
PROMPT = """What is the sum of the first 50 prime numbers?
Generate and run code for the calculation."""

response = client.models.generate_content(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

### View the generated code

The following code iterates through the response and displays any generated Python code by checking for `part.executable_code` in the response parts:

In [9]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )


```py

def is_prime(n):
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

primes = []
num = 2
while len(primes) < 50:
    if is_prime(num):
        primes.append(num)
    num += 1

sum_of_primes = sum(primes)
print(f'{sum_of_primes=}')

```


### View the code execution results

The following code iterates through the response and displays the execution result and outcome by checking for `part.code_execution_result` in the response parts:

In [10]:
for part in response.candidates[0].content.parts:
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)

`sum_of_primes=5117
`


Outcome: Outcome.OUTCOME_OK


Great! Now you have the answer (`5117`) as well as the generated (and verified via execution!) Python code.

At this point in your application, you would save the output code, result, or outcome and display it to the end-user or use it downstream in your application.

### Code execution in a chat session

This section shows how to use code execution in an interactive chat with history using the Gemini API.

You can use `client.chats.create` to create a chat session and passes in the code execution tool, enabling the model to generate and run code:

In [11]:
chat = client.chats.create(
    model=MODEL_ID,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
)

You'll start the chat by asking the model to generate sample time series data with noise and then output a sample of 10 data points:

In [12]:
PROMPT = """Create sample time series data of temperature vs. time in a test furnace.
Add noise to the data. Output a sample of 10 data points from the time series data."""

response = chat.send_message(PROMPT)

Now you can iterate through the response to display any generated Python code and execution results by checking for `part.executable_code` and `part.code_execution_result` in the response parts:

In [13]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py

import numpy as np
import pandas as pd

# Define time range (in hours)
time = np.arange(0, 10)

# Define base temperature profile (in Celsius)
temperature = np.zeros(len(time))
for i in range(len(time)):
    if time[i] <= 5:
        temperature[i] = 20 + 60 * time[i] / 5  # Linear heating from 20 to 80
    elif time[i] <= 8:
        temperature[i] = 80  # Constant temperature
    else:
        temperature[i] = 80 - 60 * (time[i] - 8) / 2  # Linear cooling from 80 to 20

# Add noise (normal distribution with mean 0 and standard deviation 5)
noise = np.random.normal(0, 5, len(time))
temperature_noisy = temperature + noise

# Create a Pandas DataFrame for better presentation
data = pd.DataFrame({'Time (hours)': time, 'Temperature (Celsius)': temperature_noisy})

# Output a sample of 10 data points
print(data.sample(10))

```


`   Time (hours)  Temperature (Celsius)
1             1              37.702092
6             6              83.020622
3             3              51.496619
7             7              77.481004
4             4              65.402154
8             8              79.815473
0             0              19.077681
2             2              48.861174
9             9              50.633805
5             5              79.577297
`


Outcome: Outcome.OUTCOME_OK


Now you can ask the model to add a smoothed data series to the time series data:

In [14]:
PROMPT = "Now add a data series that smooths the sample data."

response = chat.send_message(PROMPT)

And then display the generated Python code and execution results:

In [15]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py

import numpy as np
import pandas as pd

# Define time range (in hours)
time = np.arange(0, 10)

# Define base temperature profile (in Celsius)
temperature = np.zeros(len(time))
for i in range(len(time)):
    if time[i] <= 5:
        temperature[i] = 20 + 60 * time[i] / 5  # Linear heating from 20 to 80
    elif time[i] <= 8:
        temperature[i] = 80  # Constant temperature
    else:
        temperature[i] = 80 - 60 * (time[i] - 8) / 2  # Linear cooling from 80 to 20

# Add noise (normal distribution with mean 0 and standard deviation 5)
noise = np.random.normal(0, 5, len(time))
temperature_noisy = temperature + noise

# Calculate moving average (window size = 3)
window_size = 3
temperature_smoothed = pd.Series(temperature_noisy).rolling(window=window_size, min_periods=1).mean()

# Create a Pandas DataFrame for better presentation
data = pd.DataFrame({'Time (hours)': time, 'Temperature (Celsius)': temperature_noisy, 'Temperature Smoothed (Celsius)': temperature_smoothed})

# Output a sample of 10 data points
print(data.sample(10))

```


`   Time (hours)  Temperature (Celsius)  Temperature Smoothed (Celsius)
3             3              65.025977                       46.017601
4             4              75.764042                       61.266254
5             5              79.255920                       73.348646
0             0              24.042592                       24.042592
8             8              78.937537                       78.673018
1             1              30.018085                       27.030339
6             6              73.438907                       76.152957
7             7              83.642608                       78.779145
2             2              43.008742                       32.356473
9             9              50.521038                       71.033728
`


Outcome: Outcome.OUTCOME_OK


Finally, you can ask the model to generate descriptive statistics for the time series data:

In [16]:
PROMPT = "Now generate and output descriptive statistics on the time series data."

response = chat.send_message(PROMPT)

And then display the generated Python code and execution results:

In [17]:
for part in response.candidates[0].content.parts:
    if part.executable_code:
        display(
            Markdown(
                f"""
```py
{part.executable_code.code}
```
"""
            )
        )
    if part.code_execution_result:
        display(Markdown(f"`{part.code_execution_result.output}`"))
        print("\nOutcome:", part.code_execution_result.outcome)


```py

import numpy as np
import pandas as pd

# Define time range (in hours)
time = np.arange(0, 10)

# Define base temperature profile (in Celsius)
temperature = np.zeros(len(time))
for i in range(len(time)):
    if time[i] <= 5:
        temperature[i] = 20 + 60 * time[i] / 5  # Linear heating from 20 to 80
    elif time[i] <= 8:
        temperature[i] = 80  # Constant temperature
    else:
        temperature[i] = 80 - 60 * (time[i] - 8) / 2  # Linear cooling from 80 to 20

# Add noise (normal distribution with mean 0 and standard deviation 5)
noise = np.random.normal(0, 5, len(time))
temperature_noisy = temperature + noise

# Calculate moving average (window size = 3)
window_size = 3
temperature_smoothed = pd.Series(temperature_noisy).rolling(window=window_size, min_periods=1).mean()

# Create a Pandas DataFrame for better presentation
data = pd.DataFrame({'Time (hours)': time, 'Temperature (Celsius)': temperature_noisy, 'Temperature Smoothed (Celsius)': temperature_smoothed})

# Generate descriptive statistics
descriptive_stats = data[['Temperature (Celsius)', 'Temperature Smoothed (Celsius)']].describe()

# Output the descriptive statistics
print(descriptive_stats)

```


`       Temperature (Celsius)  Temperature Smoothed (Celsius)
count              10.000000                       10.000000
mean               57.870155                       54.802199
std                22.862738                       22.889574
min                21.859841                       21.859841
25%                38.573183                       33.414096
50%                62.191952                       59.542733
75%                76.884221                       76.168228
max                84.971472                       80.844518
`


Outcome: Outcome.OUTCOME_OK


This chat example demonstrates how you can use the Gemini API with code execution as a powerful tool for exploratory data analysis and more. Go forth and adapt this approach to your own projects and use cases!

### Code execution in a streaming session

You can also use the code execution functionality with streaming output from the Gemini API.

The following code demonstrates how the Gemini API can generate and execute code while streaming the results:

In [18]:
PROMPT = """Generate a list of 20 random names, then create a new list with just
the names containing the letter 'a', then output the number of names that
contain 'a' and finally show me that new list."""

for chunk in client.models.generate_content_stream(
    model=MODEL_ID,
    contents=PROMPT,
    config=GenerateContentConfig(
        tools=[code_execution_tool],
        temperature=0,
    ),
):
    for part in chunk.candidates[0].content.parts:
        if part.text:
            display(Markdown("#### Natural language stream"))
            display(Markdown(part.text))
            display(Markdown("---"))
        if part.executable_code:
            display(Markdown("#### Code stream"))
            display(
                Markdown(
                    f"""
```py
{part.executable_code.code}
```
"""
                )
            )
            display(Markdown("---"))
        if part.code_execution_result:
            display(Markdown("#### Code result"))
            display(
                Markdown(
                    f"""
```
{part.code_execution_result.output}
```
"""
                )
            )
            display(Markdown("---"))

#### Natural language stream

Okay

---

#### Natural language stream

, I can do that. Here's the process:

1.  Generate

---

#### Natural language stream

 a list of 20 random names. I'll use common English names for this

---

#### Natural language stream

.
2.  Create a new list containing only the names from the first list that include the letter 'a' (case-insensitive).
3.

---

#### Natural language stream

  Output the number of names in the new list.
4.  Output the new list itself.



---

#### Code stream


```py

import random

names = ['Alice', 'Bob', 'Charlie', 'David', 'Emily', 'Frank', 'Grace', 'Henry', 'Isabella', 'Jack', 'Kate', 'Liam', 'Mia', 'Noah', 'Olivia', 'Peter', 'Quinn', 'Ryan', 'Sophia', 'Thomas']

names_with_a = [name for name in names if 'a' in name.lower()]

count = len(names_with_a)

print(f'{count=}')
print(f'{names_with_a=}')

```


---

#### Code result


```
count=15
names_with_a=['Alice', 'Charlie', 'David', 'Frank', 'Grace', 'Isabella', 'Jack', 'Kate', 'Liam', 'Mia', 'Noah', 'Olivia', 'Ryan', 'Sophia', 'Thomas']

```


---

#### Natural language stream

Here

---

#### Natural language stream

's the output:

*   **Number of names containing 'a':** 

---

#### Natural language stream

15
*   **List of names containing 'a':** \['Alice', '

---

#### Natural language stream

Charlie', 'David', 'Frank', 'Grace', 'Isabella', 'Jack', 'Kate', 'Liam', 'Mia', 'Noah', 'Olivia

---

#### Natural language stream

', 'Ryan', 'Sophia', 'Thomas']

---

This streaming example demonstrated how the Gemini API can generate, execute code, and provide results within a streaming session.

## Summary

Refer to the [documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution) for more details about code execution, and in particular, the [recommendations](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/code-execution#code-execution-vs-function-calling) regarding differences between code execution and [function calling](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling).

### Next steps

- See the [Google Gen AI SDK reference docs](https://googleapis.github.io/python-genai/)
- Explore other notebooks in the [Google Cloud Generative AI GitHub repository](https://github.com/GoogleCloudPlatform/generative-ai)
- Explore AI models in [Model Garden](https://cloud.google.com/vertex-ai/generative-ai/docs/model-garden/explore-models)